arch/xtensa/core/mpu.c

```diff
@@ -0,0 +1,1235 @@
+/*
+ * Copyright (c) 2023 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <stdint.h>
+#include <string.h>
+
+#include <zephyr/kernel.h>
+#include <zephyr/spinlock.h>
+#include <zephyr/toolchain.h>
+#include <zephyr/arch/xtensa/arch_inlines.h>
+#include <zephyr/arch/xtensa/mpu.h>
+#include <zephyr/linker/linker-defs.h>
+#include <zephyr/sys/__assert.h>
+#include <zephyr/sys/util_macro.h>
+
+#include <xtensa/corebits.h>
+#include <xtensa/config/core-matmap.h>
+#include <xtensa/config/core-isa.h>
+#include <xtensa_mpu_priv.h>
+
+extern char _heap_end[];
+extern char _heap_start[];
+extern char __data_start[];
+extern char __data_end[];
+
+#ifndef CONFIG_XTENSA_MPU_ONLY_SOC_RANGES
+__weak const struct xtensa_mpu_range xtensa_soc_mpu_ranges[0];
+__weak int xtensa_soc_mpu_ranges_num;
+#endif
+
+/** MPU foreground map for kernel mode. */
+static struct xtensa_mpu_map xtensa_mpu_map_fg_kernel;
+
+/*
+ * Additional information about the MPU maps: foreground and background
+ * maps.
+ *
+ *
+ * Some things to keep in mind:
+ * - Each MPU region is described by TWO entries:
+ *   [entry_a_address, entry_b_address). For contiguous memory regions,
+ *   this should not much of an issue. However, disjoint memory regions
+ *   "waste" another entry to describe the end of those regions.
```

Comment:
One more:  this waste is significant, and seems straightforward to fix?  Just pre-compile your array of boundary entries into a "big enough" array on the stack, then pass over it slurping up duplicate boundaries that define zero-length segments.  Cheap and easy, and only has to be done once per change to the domain.  Obviously it needs some kind of handling for situations where we try to enable too many segments, but that's part of the userspace API already (and a K_OOPS() of the thread would be satisfactory handling anyway).

Reply:
I don't quite understand what you mean here. For regions of `0x1000 - 0x3000` and `0x3000 - 0x4000`, it needs 3 entries. But if you have `0x1000 - 0x2000` and `0x3000 - 0x4000`, you will need 4 entries to describe them.

(Well... for other arch's MPUs where you program both start and end addresses in the same entry, both examples only need 2 entries to describe.)

Could you describe more in details on how to pre-compile the array?

Reply:
Your initial pass might emit an array like:

```
StartAddr  Access
=========  ======
 0x1000     rw-
 0x3000     ---
 0x3000     r-x
 0x4000     ---
```

Then you just go down the initial array, and just skip any entry where the following entry has the same address (i.e. it's a definition of a zero-length segment), producing:

```
StartAddr  Access
=========  ======
 0x1000     rw-
 0x3000     r-x
 0x4000     ---
```

...which is the optimized form you want.  The same process on the disjoint second example wouldn't change the array.  The first array can be constructed cheaply on the stack with a maximally-sized (e.g. 64 entry for a 32-register MPU).  Now we can have bigger and more complicated mappings as long as lots of them are adjacent (which is reasonably common, .text is next to .rodata, kernel stacks next to user stacks, etc...)

Reply:
I can amend `consolidate_entries` to do this.

Reply:
Have amended `consolidate_entries` so if two entries have a the same address, remove the lower indexed one.